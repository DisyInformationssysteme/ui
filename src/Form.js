import React from 'react';

import RJSForm from 'react-jsonschema-form';

import Button from 'react-bootstrap/lib/Button';

import SchemaField from './SchemaField';
import RadioOrSelectWidget from './widgets/RadioOrSelectWidget';

const customWidgets = {
	radioOrSelectWidget: RadioOrSelectWidget,
};

const customUiSchema = {
	'ui:widget': 'radioOrSelectWidget',
};

class Form extends React.Component {

	constructor(props) {
		super(props);
		this.handleSchemaChange = this.handleSchemaChange.bind(this);
	}

	handleSchemaChange(change) {
		// TODO: Handle Schema Change
		return this.props.onChange && this.props.onChange(change);
	}

	render() {
		const schema = (this.props.data && this.props.data.jsonSchema) ||
			this.props.jsonSchema || this.props.schema;
		const uiSchema = {
			...((this.props.data && this.props.data.uiSchema) || this.props.uiSchema),
			...customUiSchema,
		};
		const formData = (this.props.data && this.props.data.properties) || this.props.formData;
		const fields = { SchemaField };
		const actions = this.props.actions ? this.props.actions.map((action, index) => (
			<Button
				key={index}
				bsStyle={action.style}
				type={action.type}
				onClick={action.onClick}
				title={action.title}
			>
				{action.icon ? <i className={action.icon} /> : null }
				{action.label}
			</Button>
		)) : <Button bsStyle="primary" type="submit">Submit</Button>;
		return (
			<RJSForm
				{...this.props}
				schema={schema}
				uiSchema={uiSchema}
				formData={formData}
				fields={fields}
				widgets={customWidgets}
				onChange={this.handleSchemaChange}
			>
				{actions}
			</RJSForm>
		);
	}
}

Form.propTypes = {
	data: React.PropTypes.shape({
		jsonSchema: React.PropTypes.object,
		uiSchema: React.PropTypes.object,
		properties: React.PropTypes.object,
		formData: React.PropTypes.object,
	}),
	jsonSchema: React.PropTypes.object,
	uiSchema: React.PropTypes.object,
	formData: React.PropTypes.object,
	schema: React.PropTypes.object,
	onChange: React.PropTypes.func,
	actions: React.PropTypes.arrayOf(React.PropTypes.shape({
		style: React.PropTypes.string,
		type: React.PropTypes.oneOf(['submit', 'reset', 'button']),
		onClick: React.PropTypes.func,
		label: React.PropTypes.string,
		icon: React.PropTypes.string,
		title: React.PropTypes.string,
	})),
};

export default Form;
