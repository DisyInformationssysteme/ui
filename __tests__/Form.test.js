import React from 'react';
import { mount } from 'enzyme';
import {
  FormControl,
  Checkbox,
} from 'react-bootstrap';
import Form from '../src/Form';

const schema = {
	title: 'TestForm',
	type: 'object',
	properties: {
		name: {
			type: 'string',
			title: 'Name',
			default: 'John Doe',
		},
		admin: {
			type: 'boolean',
			title: 'Admin',
			default: true,
		},
		roles: {
			type: 'string',
			enum: ['dev', 'pm'],
			enumName: ['Developer', 'Project Manager'],
		},
	},
};

describe('<Form/>', () => {
	let wrapper;
	const onSubmit = jest.fn();
	const onChange = jest.fn();

	beforeEach(() => {
		wrapper = mount(<Form schema={schema} onSubmit={onSubmit} onChange={onChange} />);
	});

	it('Renders the <FormControl/> component', () => {
		expect(wrapper.containsMatchingElement(<FormControl />)).toBe(true);
	});
	it('Renders the <Checkbox/> component', () => {
		expect(wrapper.containsMatchingElement(<Checkbox />)).toBe(true);
	});

	it('Renders the <FormControl/> component', () => {
		expect(wrapper.containsMatchingElement(<FormControl />)).toBe(true);
	});

	// TODO: Follow what's happeing on https://github.com/airbnb/enzyme/issues/364
	// and update accordingly.
	// So far it's not possible to get the onChange method to be bubbled up to the
	// form
	xit('Handles changes', () => {
		const input = wrapper.find('input').first();
		input.simulate('change', { target: { value: 'Test' } });
		expect(input.props().value).toEqual('Test');
	// expect(onChange.mock.calls.length).toEqual(1);
	});
	it('Handles submit', () => {
		wrapper.simulate('submit');
		expect(onSubmit.mock.calls.length).toEqual(1);
	});
});
