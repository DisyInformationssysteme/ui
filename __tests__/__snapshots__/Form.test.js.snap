exports[`<Form/> <Form actions/> Renders form with custom css 1`] = `
<form
  acceptCharset={undefined}
  action={undefined}
  autoComplete={undefined}
  className="form"
  encType={undefined}
  id={undefined}
  method={undefined}
  name={undefined}
  onSubmit={[Function]}
  target={undefined}>
  <div>
    <fieldset>
      <legend
        id="root__title">
        TestForm
      </legend>
      <div>
        <div
          className={undefined}
          style={
            Object {
              "backgroundColor": "transparent",
              "cursor": "auto",
              "display": "inline-block",
              "fontFamily": "Roboto, sans-serif",
              "fontSize": 16,
              "height": 72,
              "lineHeight": "24px",
              "muiPrepared": true,
              "position": "relative",
              "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": 256,
            }
          }>
          <label
            className={undefined}
            htmlFor="root_name"
            onTouchTap={undefined}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.3)",
                "lineHeight": "22px",
                "muiPrepared": true,
                "pointerEvents": "none",
                "position": "absolute",
                "top": 38,
                "transform": "scale(0.75) translate(0, -28px)",
                "transformOrigin": "left top",
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "userSelect": "none",
                "zIndex": 1,
              }
            }>
            Name
          </label>
          <input
            autoFocus={false}
            disabled={false}
            id="root_name"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder=""
            readOnly={false}
            required={false}
            style={
              Object {
                "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                "appearance": "textfield",
                "backgroundColor": "rgba(0,0,0,0)",
                "border": "none",
                "boxSizing": "border-box",
                "color": "#000000",
                "cursor": "inherit",
                "font": "inherit",
                "height": "100%",
                "marginTop": 14,
                "muiPrepared": true,
                "outline": "none",
                "padding": 0,
                "position": "relative",
                "width": "100%",
              }
            }
            type="text"
            value="John Doe" />
          <div>
            <hr
              style={
                Object {
                  "border": "none",
                  "borderBottom": "solid 1px",
                  "borderColor": "#e0e0e0",
                  "bottom": 8,
                  "boxSizing": "content-box",
                  "margin": 0,
                  "muiPrepared": true,
                  "position": "absolute",
                  "width": "100%",
                }
              } />
            <hr
              style={
                Object {
                  "border": "none",
                  "borderBottom": "solid 2px",
                  "borderColor": "rgba(0, 0, 0, 0.54)",
                  "bottom": 8,
                  "boxSizing": "content-box",
                  "margin": 0,
                  "muiPrepared": true,
                  "position": "absolute",
                  "transform": "scaleX(0)",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "width": "100%",
                }
              } />
          </div>
        </div>
        <div />
        <div />
      </div>
      <div>
        <div
          autoFocus={false}
          className={undefined}
          multiple={undefined}
          readOnly={false}
          required={false}
          style={
            Object {
              "backgroundColor": "transparent",
              "cursor": "auto",
              "display": "inline-block",
              "fontFamily": "Roboto, sans-serif",
              "fontSize": 16,
              "height": 72,
              "lineHeight": "24px",
              "muiPrepared": true,
              "position": "relative",
              "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "width": 256,
            }
          }>
          <label
            className={undefined}
            htmlFor="root_roles"
            onTouchTap={undefined}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.3)",
                "lineHeight": "22px",
                "muiPrepared": true,
                "pointerEvents": "none",
                "position": "absolute",
                "top": 38,
                "transform": "scale(0.75) translate(0, -28px)",
                "transformOrigin": "left top",
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "userSelect": "none",
                "zIndex": 1,
              }
            }>
            roles
          </label>
          <div
            className={undefined}
            disabled={false}
            id="root_roles"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            style={
              Object {
                "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                "backgroundColor": "rgba(0,0,0,0)",
                "border": "none",
                "boxSizing": "border-box",
                "color": "#000000",
                "cursor": "inherit",
                "display": "block",
                "font": "inherit",
                "fontFamily": "Roboto, sans-serif",
                "fontSize": 15,
                "height": "100%",
                "marginTop": 14,
                "muiPrepared": true,
                "outline": "none",
                "padding": 0,
                "position": "relative",
                "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": "100%",
              }
            }>
            <div
              onTouchTap={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                  "height": "100%",
                  "muiPrepared": true,
                  "position": "relative",
                  "width": "100%",
                }
              }>
              <div
                style={
                  Object {
                    "boxSizing": "border-box",
                    "content": "\' \'",
                    "display": "table",
                    "muiPrepared": true,
                  }
                } />
              <div
                style={
                  Object {
                    "color": "#000000",
                    "lineHeight": "56px",
                    "muiPrepared": true,
                    "opacity": 1,
                    "paddingLeft": 0,
                    "paddingRight": 48,
                    "position": "relative",
                    "top": 6,
                  }
                }>
                dev
              </div>
              <svg
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "color": "#000000",
                    "display": "inline-block",
                    "fill": "#e0e0e0",
                    "height": 24,
                    "muiPrepared": true,
                    "position": "absolute",
                    "right": 0,
                    "top": 22,
                    "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    "userSelect": "none",
                    "width": 24,
                  }
                }
                viewBox="0 0 24 24">
                <path
                  d="M7 10l5 5 5-5z" />
              </svg>
              <div
                style={
                  Object {
                    "borderTop": "none",
                    "bottom": 1,
                    "left": 0,
                    "margin": "-1px 24px",
                    "muiPrepared": true,
                    "position": "absolute",
                    "right": 0,
                  }
                } />
              <div
                style={
                  Object {
                    "boxSizing": "border-box",
                    "clear": "both",
                    "content": "\' \'",
                    "display": "table",
                    "muiPrepared": true,
                  }
                } />
            </div>
            <div />
          </div>
          <div>
            <hr
              style={
                Object {
                  "border": "none",
                  "borderBottom": "solid 1px",
                  "borderColor": "#e0e0e0",
                  "bottom": 8,
                  "boxSizing": "content-box",
                  "margin": 0,
                  "muiPrepared": true,
                  "position": "absolute",
                  "width": "100%",
                }
              } />
            <hr
              style={
                Object {
                  "border": "none",
                  "borderBottom": "solid 2px",
                  "borderColor": "rgba(0, 0, 0, 0.54)",
                  "bottom": 8,
                  "boxSizing": "content-box",
                  "margin": 0,
                  "muiPrepared": true,
                  "position": "absolute",
                  "transform": "scaleX(0)",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "width": "100%",
                }
              } />
          </div>
        </div>
        <div />
        <div />
      </div>
    </fieldset>
    <div />
    <div />
  </div>
  <div
    className="buttons">
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      title={undefined}
      type="submit">
      Submit
    </button>
    <button
      className="btn btn-link"
      disabled={false}
      onClick={[Function]}
      title={undefined}
      type="reset">
      Reset
    </button>
  </div>
</form>
`;
